libflux: msg API cleanup

chu11 opened the pull request:
Various cleanup while I was working on #3617

garlick commented on src/common/libflux/message.c:
```diff
@@ -277,11 +277,15 @@ void *flux_msg_aux_get (const flux_msg_t *msg, const char *name)
 }
 
-size_t flux_msg_encode_size (const flux_msg_t *msg)
+ssize_t flux_msg_encode_size (const flux_msg_t *msg)
```
Probably should update the one user in librouter/sendfd.c to handle the error case that's now possible.

garlick commented on src/common/libflux/message.c:
```diff
@@ -1683,11 +1683,6 @@ flux_msg_t *flux_msg_recvzsock (void *sock)
     return msg;
 }
 
-int flux_msg_frames (const flux_msg_t *msg)
-{
-    return zmsg_size (msg->zmsg);
-}
-
```
I think this is here mainly for testing.  By removing the frame count checks from the test, are we diminishing our unit test?  Since we will still need to have a variable length stack of message frames in the rework, does it make sense to either have a function that returns a count of those, or just keep this function?  There will still be the same message frames on the wire in the rework unless we are planning to redo the protocol and RFC 3.

See zmq_send(3) - the terminology is slightly different without czmq.  zmq refers to multi-part messages, i.e. "message parts" instead of "message frames".  So maybe a rename to `flux_msg_part_count()` or similar if we decide to keep this function.

chu11 replied:
We could just keep the function.  I was mostly thinking that counting frames suggests how data is stored internally and since we're planning on changing that, the function would have less purpose going forward.

garlick replied:
I think you make a good case for not exposing this in the public API.  If you plan to have a `message_private.h` in the next PR to privately share the message struct with librouter, maybe this function belongs there under a new name like `msg_multipart_count()` or similar.

garlick reviewed (commented):
It would be good to give each of them a "problem style" commit message.

chu11 commented:
rebased & re-pushed.  updated more comments to have problem/solution format.  checked return code in `sendfd` per comment above.  Removed commit that removed `flux_msg_frames()`.

garlick commented on src/common/libflux/message.c:
```diff
@@ -419,7 +419,7 @@ int flux_msg_get_flags (const flux_msg_t *msg, uint8_t *fl)
 
 int flux_msg_set_private (flux_msg_t *msg)
 {
-    uint8_t flags;
```
Just a tiny bit more info in the commit message explaining why this is a problem now, as opposed to before this PR(?) would be helpful.

chu11 replied:
this fix was "preemptive" as it really appeared after some refactoring for #3617.  But I'll add a note of how it is "sensitive" to changes.

garlick replied:
When I'm perusing C code, I usually take a local variable initialization as a subconscious hint to watch out for a code path in which its value is used before set, so when there isn't one, it tends to make my brain do extra work.  For that reason, sometimes I'll clean up a function by _removing_ unnecessary initialization when it's obvious it's not needed.  I'm not sure if that's reasonable, or if I should reprogram myself :-)

chu11 replied:
i'm actually the same way and am not entirely sure why this started popping up recently.  It's this style of code that was hit:

```
int flags;
if (get_flags (ctx, &flags) < 0)
   goto error;
```

i don't know why changes somehow triggered the warning :-(   Perhaps some other code changes led to an extra pass in the compiler and that's all that happened.

garlick reviewed (approved):
Looks good!  Just one nit on a commit message.  Feel free to fix or not, and force push and set MWP when done.

Thanks for cleaning this up!

garlick commented:
Did you mean to close this and delete the branch without merging?

chu11 commented:
oh crap!  I was cleaning up my remote branches, guess i deleted the wrong one.

chu11 commented:
crap, i guess i can't re-open a PR, I'll have to re-open this.  my bad
